### PR TITLE
docs: Add replay diagram and notes to checkpoints and ReplayPlanner

### PR DESCRIPTION
**Pull request by alamb**
# Rationale
* Closes https://github.com/influxdata/influxdb_iox/issues/2166 by writing down what @tustvold and @crepareum told me this morning

# Changes:
* Add docstrings and ASCII art

# ASCII Art
Here is the mono draw file, should you be into that sort of thing (aka if I am ever looking for this in the future): 
[sequence_numbers.zip](https://github.com/influxdata/influxdb_iox/files/6926600/sequence_numbers.zip)

### CODE REVIEW

File: persistence_windows/src/checkpoint.rs [hunk 1]
```diff
@@ -1,9 +1,65 @@
...
+//! *max sequence number*: any data with sequence number higher than
+//! this number had not yet been seen by IOx when the checkpoint was
+//! persisted to object store
+//!
+//! Thus from the above diagram we can see:
+//!
+//! 1. *min sequence number* is 100, meaning that all data with
+//! sequence number < 100 is persisted in object store
+//!
+//! 2. *max sequence number* is 103, meaning there is definitely no no
```
Comment (tustvold) [inline persistence_windows/src/checkpoint.rs:56]:
```suggestion
//! 2. *max sequence number* is 103, meaning there is definitely no
```

File: persistence_windows/src/checkpoint.rs [hunk 2]
```diff
@@ -1,9 +1,65 @@
...
+//! *min sequence number*: all data with sequence number lower than
+//! this have been persisted to object store
+//!
+//! *max sequence number*: any data with sequence number higher than
+//! this number had not yet been seen by IOx when the checkpoint was
+//! persisted to object store
+//!
+//! Thus from the above diagram we can see:
+//!
+//! 1. *min sequence number* is 100, meaning that all data with
```
Comment (tustvold) [inline persistence_windows/src/checkpoint.rs:53]: There is an edge case when all data is persisted, in such a case the min sequence number will be None implying all data up to and including the max sequence number has been persisted

File: persistence_windows/src/checkpoint.rs [hunk 3]
```diff
@@ -277,7 +358,8 @@ pub struct PartitionCheckpoint {
     /// Maps `sequencer_id` to the to-be-persisted minimum and seen maximum sequence numbers.
     sequencer_numbers: BTreeMap<u32, OptionalMinMaxSequence>,
 
-    /// Minimum unpersisted timestamp.
+    /// Minimum unpersisted timestamp value of the TIMESTAMP_COLUMN
+    /// (aka timestamps in the data, aka "flush timestamp")
```
Comment (tustvold) [inline persistence_windows/src/checkpoint.rs:362]: This currently is actually "flush timestamp" + 1 nanosecond

File: persistence_windows/src/checkpoint.rs [hunk 4]
```diff
@@ -265,7 +321,32 @@ pub enum Error {
...
+///
+/// The min_unpersisted timestamp is a value relative to the data's
+/// value in the TIMESTAMP_COLUMN. The min/max sequence numbers are
+/// relative to their respective sequencers.
+///
+/// Invariants at the time this structure was written:
+///
+/// ```text
+/// ┌───────────────────┬ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─┌─────────────────────┐
+/// │seq < min_sequence │ time <= min_unpersisted  │ seq > max sequence  │
```
Comment (tustvold) [inline persistence_windows/src/checkpoint.rs:339]: As mentioned below, currently the formulation is min_unpersisted not flush timestamp and so equality implies unpersisted. This is a historical artifact we plan to fix.

File: persistence_windows/src/checkpoint.rs [hunk 5]
```diff
@@ -265,7 +321,32 @@ pub enum Error {
...
+///
+/// This structure contains the minimum and maximum sequence numbers
+/// for each sequencer for each partition along with the
+/// min_unpersisted timestamp ("flush timestamp").
+///
+/// The min_unpersisted timestamp is a value relative to the data's
+/// value in the TIMESTAMP_COLUMN. The min/max sequence numbers are
+/// relative to their respective sequencers.
+///
+/// Invariants at the time this structure was written:
```
Comment (tustvold) [inline persistence_windows/src/checkpoint.rs:335]: Given this is described as being at the time a "Persist" operation was planned, this should possibly mention the fact that it describes the state of the system after the planned operation completes

**Review by tustvold — APPROVED** [review]
Thank you for doing this, looks good to me, only some minor points

**Review by crepererum — APPROVED** [review]
minor nits, thanks for the write up!

File: persistence_windows/src/checkpoint.rs [hunk 6]
```diff
@@ -1,9 +1,65 @@
...
+//!
+//!                     min sequence     max sequence
+//!                           │                │
+//!    Input Stream       ┌───┘          ┌─────┘
+//!   from Sequencer      ▼              ▼
+//!               ┌────┬────┬────┬────┬────┬────┬────┬────┬────┬────┐
+//!               │Data│Data│Data│Data│Data│Data│Data│Data│Data│Data│
+//!               └────┴────┴────┴────┴────┴────┴────┴────┴────┴────┘
+//!                99  100   101  102  103  104  105  106  107  108     Sequence Number
+//!                                                    ▲
```
[the remark this answers is not in the document]
Comment (alamb) [reply persistence_windows/src/checkpoint.rs:27]: My rationale for including the IOx node was to illustrate the "system level" idea that data was being read from the input stream and being written to parquet. I will point it out in the text as you suggest.

File: persistence_windows/src/checkpoint.rs (hunk 2 above)
Comment (alamb) [reply persistence_windows/src/checkpoint.rs:56]: I added a comment to that effect on `OptionalMinMaxSequence`. Is this sentence still logically accurate (aka a min_sequence_number of `None` implies min_sequence_number == max_sequence_number)?

File: persistence_windows/src/checkpoint.rs (hunk 4 above)
Comment (alamb) [reply persistence_windows/src/checkpoint.rs]: will change bounds

File: persistence_windows/src/checkpoint.rs [hunk 7]
```diff
@@ -335,9 +417,29 @@ impl PartitionCheckpoint {
...
+/// the appropriate min and max sequences are considered. Data prior
+/// to the min sequence is already persisted and data after the max is
+/// known to be unpersisted.
+///
+/// ```text
+/// ┌───────────────────┬ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─┌─────────────────────┐
+/// │                   │  seq >= min_sequence &&  │                     │
+/// │seq < min_sequence │   seq <= max_sequence    │ seq > max sequence  │
+/// │                   │                          │                     │
+/// │     PERSISTED     │     MAYBE PERSISTED      │     UNPERSISTED     │
```
[the remark this answers is not in the document]
Comment (alamb) [reply persistence_windows/src/checkpoint.rs:447]: That is a good subtlety -- I have added it

File: persistence_windows/src/checkpoint.rs [hunk 8]
```diff
@@ -448,11 +550,26 @@ impl PersistCheckpointBuilder {
...
 }
 
-/// Plan your sequencer replays after server restarts.
+/// Plans Replays after database restarts.
+///
+/// Database Replay is the reading of data from the sequencers that
+/// may have been partially persisted. The goals of replay are:
+///
+/// 1. Ensure that all data is available for query up to the
+/// sequencer's max sequence number in the DatabaseCheckpoint by
```
[the remark this answers is not in the document]
Comment (alamb) [reply persistence_windows/src/checkpoint.rs]: I reworded this to remove the reference to `DatabaseCheckpoint` as you suggested:

```rust
/// 1. Ensure that all data is available for query up to the largest
/// sequence number that had been read at the time of the most recent
/// persist operation, by re-ingesting necessary data from each
/// sequencer.
```

**Pull request merged**